core/app/models/spree/promotion_handler/free_shipping.rb

```diff
@@ -18,7 +18,7 @@ def activate
         end
 
         not_connected_automatic_promotions.each do |promotion|
-          promotion.activate(order: order)
+          promotion.activate(order: order) if promotion.eligible?(order)
```

Comment:
Would you mind making this a normal `if` instead of a trailing `if`?  The condition is quite important, as you've pointed out, so it'd be nice to make sure it stands out